// http-mocks/esm2015/lib/utils/delay.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/utils/delay.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// SOURCE: https://github.com/sindresorhus/delay
/**
 * @record
 */
function AbortSignal() { }
if (false) {
    /** @type {?} */
    AbortSignal.prototype.aborted;
    /**
     * @param {?} type
     * @param {?} listener
     * @param {?=} options
     * @return {?}
     */
    AbortSignal.prototype.addEventListener = function (type, listener, options) { };
    /**
     * @param {?} type
     * @param {?} listener
     * @return {?}
     */
    AbortSignal.prototype.removeEventListener = function (type, listener) { };
}
/**
 * @record
 */
function Options() { }
if (false) {
    /** @type {?|undefined} */
    Options.prototype.signal;
}
/**
 * @record
 * @template T
 */
function ClearablePromise() { }
if (false) {
    /**
     * @return {?}
     */
    ClearablePromise.prototype.clear = function () { };
}
/**
 * @record
 */
function Delay() { }
if (false) {
    /** @type {?|undefined} */
    Delay.prototype.reject;
    /** @type {?|undefined} */
    Delay.prototype.createWithTimers;
    /* Skipping unhandled member: (milliseconds: number, options?: Options): ClearablePromise<void>;*/
    /* Skipping unhandled member: <T>(
        milliseconds: number,
        options?: Options & {
          value: T;
        }
      ): ClearablePromise<T>;*/
}
/** @type {?} */
const createAbortError = (/**
 * @return {?}
 */
() => {
    /** @type {?} */
    const error = new Error('Delay aborted');
    error.name = 'AbortError';
    return error;
});
const ɵ0 = createAbortError;
/** @type {?} */
export const createDelay = (/**
 * @param {?} __0
 * @return {?}
 */
({ clearTimeout: defaultClear, setTimeout: set, willResolve }) => (/**
 * @param {?} ms
 * @param {?=} __1
 * @return {?}
 */
(ms, { value, signal } = (/** @type {?} */ ({}))) => {
    if (signal && signal.aborted) {
        return Promise.reject(createAbortError());
    }
    /** @type {?} */
    let timeoutId;
    /** @type {?} */
    let settle;
    /** @type {?} */
    let rejectFn;
    /** @type {?} */
    const clear = defaultClear || clearTimeout;
    /** @type {?} */
    const signalListener = (/**
     * @return {?}
     */
    () => {
        clear(timeoutId);
        rejectFn(createAbortError());
    });
    /** @type {?} */
    const cleanup = (/**
     * @return {?}
     */
    () => {
        if (signal) {
            signal.removeEventListener('abort', signalListener);
        }
    });
    /** @type {?} */
    const delayPromise = new Promise((/**
     * @param {?} resolve
     * @param {?} reject
     * @return {?}
     */
    (resolve, reject) => {
        settle = (/**
         * @return {?}
         */
        () => {
            cleanup();
            if (willResolve) {
                resolve(value);
            }
            else {
                reject(value);
            }
        });
        rejectFn = reject;
        timeoutId = (set || setTimeout)(settle, ms);
    }));
    if (signal) {
        signal.addEventListener('abort', signalListener, { once: true });
    }
    delayPromise.clear = (/**
     * @return {?}
     */
    () => {
        clear(timeoutId);
        timeoutId = null;
        settle();
    });
    return delayPromise;
}));
/** @type {?} */
export const delay = createDelay({ willResolve: true });
delay.reject = createDelay({ willResolve: false });
delay.createWithTimers = (/**
 * @param {?} __0
 * @return {?}
 */
({ clearTimeout, setTimeout }) => {
    /** @type {?} */
    const innerDelay = createDelay({
        clearTimeout,
        setTimeout,
        willResolve: true
    });
    innerDelay.reject = createDelay({
        clearTimeout,
        setTimeout,
        willResolve: false
    });
    return innerDelay;
});
export { ɵ0 };
