// http-mocks/esm2015/lib/utils/fetch.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/utils/fetch.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import fetchMock from 'fetch-mock/es5/client';
import { parseQueryParams } from './query';
import { logMock } from './log';
import { handleProxyFn } from './proxy';
/** @type {?} */
export const fallbackToNetworkFetch = (/**
 * @param {?} fallbackToNetwork
 * @return {?}
 */
(fallbackToNetwork) => {
    fetchMock.config.fallbackToNetwork = fallbackToNetwork;
});
/** @type {?} */
export const overwriteRoutesFetch = (/**
 * @param {?} overwriteRoutes
 * @return {?}
 */
(overwriteRoutes) => {
    fetchMock.config.overwriteRoutes = overwriteRoutes;
});
/** @type {?} */
export const createFetchMock = (/**
 * @param {?} mock
 * @param {?} loggingEnabled
 * @param {?} responseProxyFn
 * @return {?}
 */
(mock, loggingEnabled, responseProxyFn) => {
    fetchMock.mock(mock.url, (/**
     * @param {?} url
     * @param {?} fetchMockOptions
     * @return {?}
     */
    (url, fetchMockOptions) => {
        var _a, _b;
        /** @type {?} */
        const body = ((_a = fetchMockOptions) === null || _a === void 0 ? void 0 : _a.body) || {};
        /** @type {?} */
        const headers = ((_b = fetchMockOptions) === null || _b === void 0 ? void 0 : _b.headers) || {};
        /** @type {?} */
        const createdURL = createURL(url);
        // request
        /** @type {?} */
        const requestQuery = parseQueryParams(createdURL !== null
            ? createdURL.search
            : removeAllBeforeChar(url.toString(), '?'));
        /** @type {?} */
        const requestBody = typeof body === 'string' ? JSON.parse(body) : body;
        /** @type {?} */
        const requestHeaders = headers;
        // response
        /** @type {?} */
        const responseBody = mock.responseFn(requestQuery, requestBody);
        /** @type {?} */
        const responseData = handleProxyFn(responseBody, mock, {
            requestQuery,
            requestBody,
            requestHeaders
        }, responseProxyFn);
        /** @type {?} */
        const response = {
            headers: responseData.responseHeaders,
            status: responseData.responseCode,
            body: responseData.responseBody
        };
        if (loggingEnabled) {
            logMock(mock, {
                requestQuery,
                requestBody,
                requestHeaders
            }, responseData);
        }
        return response;
    }), {
        delay: mock.delay
    });
});
/**
 * Tries to create a new instace of URL()
 * @type {?}
 */
const createURL = (/**
 * @param {?} url
 * @return {?}
 */
(url) => {
    try {
        return new URL(`${url}`);
    }
    catch (_) {
        return null;
    }
});
const ɵ0 = createURL;
/**
 * Removes all chars before specific char (char includes).
 * If char doesn't exist, returns an empty string.
 * @type {?}
 */
const removeAllBeforeChar = (/**
 * @param {?} value
 * @param {?} char
 * @return {?}
 */
(value, char) => {
    if (!value.includes(char)) {
        return '';
    }
    return `${value}`.split(char).pop();
});
const ɵ1 = removeAllBeforeChar;
export { ɵ0, ɵ1 };
