// http-mocks/esm5/lib/utils/delay.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/utils/delay.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// SOURCE: https://github.com/sindresorhus/delay
/**
 * @record
 */
function AbortSignal() { }
if (false) {
    /** @type {?} */
    AbortSignal.prototype.aborted;
    /**
     * @param {?} type
     * @param {?} listener
     * @param {?=} options
     * @return {?}
     */
    AbortSignal.prototype.addEventListener = function (type, listener, options) { };
    /**
     * @param {?} type
     * @param {?} listener
     * @return {?}
     */
    AbortSignal.prototype.removeEventListener = function (type, listener) { };
}
/**
 * @record
 */
function Options() { }
if (false) {
    /** @type {?|undefined} */
    Options.prototype.signal;
}
/**
 * @record
 * @template T
 */
function ClearablePromise() { }
if (false) {
    /**
     * @return {?}
     */
    ClearablePromise.prototype.clear = function () { };
}
/**
 * @record
 */
function Delay() { }
if (false) {
    /** @type {?|undefined} */
    Delay.prototype.reject;
    /** @type {?|undefined} */
    Delay.prototype.createWithTimers;
    /* Skipping unhandled member: (milliseconds: number, options?: Options): ClearablePromise<void>;*/
    /* Skipping unhandled member: <T>(
        milliseconds: number,
        options?: Options & {
          value: T;
        }
      ): ClearablePromise<T>;*/
}
/** @type {?} */
var createAbortError = (/**
 * @return {?}
 */
function () {
    /** @type {?} */
    var error = new Error('Delay aborted');
    error.name = 'AbortError';
    return error;
});
var ɵ0 = createAbortError;
/** @type {?} */
export var createDelay = (/**
 * @param {?} __0
 * @return {?}
 */
function (_a) {
    var defaultClear = _a.clearTimeout, set = _a.setTimeout, willResolve = _a.willResolve;
    return (/**
     * @param {?} ms
     * @param {?=} __1
     * @return {?}
     */
    function (ms, _a) {
        var _b = _a === void 0 ? (/** @type {?} */ ({})) : _a, value = _b.value, signal = _b.signal;
        if (signal && signal.aborted) {
            return Promise.reject(createAbortError());
        }
        /** @type {?} */
        var timeoutId;
        /** @type {?} */
        var settle;
        /** @type {?} */
        var rejectFn;
        /** @type {?} */
        var clear = defaultClear || clearTimeout;
        /** @type {?} */
        var signalListener = (/**
         * @return {?}
         */
        function () {
            clear(timeoutId);
            rejectFn(createAbortError());
        });
        /** @type {?} */
        var cleanup = (/**
         * @return {?}
         */
        function () {
            if (signal) {
                signal.removeEventListener('abort', signalListener);
            }
        });
        /** @type {?} */
        var delayPromise = new Promise((/**
         * @param {?} resolve
         * @param {?} reject
         * @return {?}
         */
        function (resolve, reject) {
            settle = (/**
             * @return {?}
             */
            function () {
                cleanup();
                if (willResolve) {
                    resolve(value);
                }
                else {
                    reject(value);
                }
            });
            rejectFn = reject;
            timeoutId = (set || setTimeout)(settle, ms);
        }));
        if (signal) {
            signal.addEventListener('abort', signalListener, { once: true });
        }
        delayPromise.clear = (/**
         * @return {?}
         */
        function () {
            clear(timeoutId);
            timeoutId = null;
            settle();
        });
        return delayPromise;
    });
});
/** @type {?} */
export var delay = createDelay({ willResolve: true });
delay.reject = createDelay({ willResolve: false });
delay.createWithTimers = (/**
 * @param {?} __0
 * @return {?}
 */
function (_a) {
    var clearTimeout = _a.clearTimeout, setTimeout = _a.setTimeout;
    /** @type {?} */
    var innerDelay = createDelay({
        clearTimeout: clearTimeout,
        setTimeout: setTimeout,
        willResolve: true
    });
    innerDelay.reject = createDelay({
        clearTimeout: clearTimeout,
        setTimeout: setTimeout,
        willResolve: false
    });
    return innerDelay;
});
export { ɵ0 };
