// http-mocks/esm5/lib/utils/fetch.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/utils/fetch.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import fetchMock from 'fetch-mock/es5/client';
import { parseQueryParams } from './query';
import { logMock } from './log';
import { handleProxyFn } from './proxy';
/** @type {?} */
export var fallbackToNetworkFetch = (/**
 * @param {?} fallbackToNetwork
 * @return {?}
 */
function (fallbackToNetwork) {
    fetchMock.config.fallbackToNetwork = fallbackToNetwork;
});
/** @type {?} */
export var overwriteRoutesFetch = (/**
 * @param {?} overwriteRoutes
 * @return {?}
 */
function (overwriteRoutes) {
    fetchMock.config.overwriteRoutes = overwriteRoutes;
});
/** @type {?} */
export var createFetchMock = (/**
 * @param {?} mock
 * @param {?} loggingEnabled
 * @param {?} responseProxyFn
 * @return {?}
 */
function (mock, loggingEnabled, responseProxyFn) {
    fetchMock.mock(mock.url, (/**
     * @param {?} url
     * @param {?} fetchMockOptions
     * @return {?}
     */
    function (url, fetchMockOptions) {
        var _a, _b;
        /** @type {?} */
        var body = ((_a = fetchMockOptions) === null || _a === void 0 ? void 0 : _a.body) || {};
        /** @type {?} */
        var headers = ((_b = fetchMockOptions) === null || _b === void 0 ? void 0 : _b.headers) || {};
        /** @type {?} */
        var createdURL = createURL(url);
        // request
        /** @type {?} */
        var requestQuery = parseQueryParams(createdURL !== null
            ? createdURL.search
            : removeAllBeforeChar(url.toString(), '?'));
        /** @type {?} */
        var requestBody = typeof body === 'string' ? JSON.parse(body) : body;
        /** @type {?} */
        var requestHeaders = headers;
        // response
        /** @type {?} */
        var responseBody = mock.responseFn(requestQuery, requestBody);
        /** @type {?} */
        var responseData = handleProxyFn(responseBody, mock, {
            requestQuery: requestQuery,
            requestBody: requestBody,
            requestHeaders: requestHeaders
        }, responseProxyFn);
        /** @type {?} */
        var response = {
            headers: responseData.responseHeaders,
            status: responseData.responseCode,
            body: responseData.responseBody
        };
        if (loggingEnabled) {
            logMock(mock, {
                requestQuery: requestQuery,
                requestBody: requestBody,
                requestHeaders: requestHeaders
            }, responseData);
        }
        return response;
    }), {
        delay: mock.delay
    });
});
/**
 * Tries to create a new instace of URL()
 * @type {?}
 */
var createURL = (/**
 * @param {?} url
 * @return {?}
 */
function (url) {
    try {
        return new URL("" + url);
    }
    catch (_) {
        return null;
    }
});
var ɵ0 = createURL;
/**
 * Removes all chars before specific char (char includes).
 * If char doesn't exist, returns an empty string.
 * @type {?}
 */
var removeAllBeforeChar = (/**
 * @param {?} value
 * @param {?} char
 * @return {?}
 */
function (value, char) {
    if (!value.includes(char)) {
        return '';
    }
    return ("" + value).split(char).pop();
});
var ɵ1 = removeAllBeforeChar;
export { ɵ0, ɵ1 };
