// http-mocks/esm5/lib/utils/xhr.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/utils/xhr.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { __awaiter, __generator } from "tslib";
import xhrMock, { proxy } from 'xhr-mock';
import { delay } from './delay';
import { logMock } from './log';
import { handleProxyFn } from './proxy';
/** @type {?} */
export var setupXHRMock = (/**
 * @return {?}
 */
function () {
    xhrMock.setup();
});
/** @type {?} */
export var fallbackToNetworkXHR = (/**
 * @param {?} fallbackToNetwork
 * @return {?}
 */
function (fallbackToNetwork) {
    if (fallbackToNetwork) {
        xhrMock.use(proxy);
    }
});
/** @type {?} */
export var createXHRMock = (/**
 * @param {?} mock
 * @param {?} loggingEnabled
 * @param {?=} responseProxyFn
 * @return {?}
 */
function (mock, loggingEnabled, responseProxyFn) {
    xhrMock.use(mock.method, mock.url, (/**
     * @param {?} mockRequest
     * @param {?} mockResponse
     * @return {?}
     */
    function (mockRequest, mockResponse) { return __awaiter(void 0, void 0, void 0, function () {
        var requestQuery, requestBody, requestHeaders, responseBody, responseData, response;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    // request
                    requestQuery = mockRequest.url().query || {};
                    requestBody = mockRequest.body() != null ? JSON.parse(mockRequest.body()) : {};
                    requestHeaders = mockRequest.headers();
                    // response
                    responseBody = mock.responseFn(requestQuery, requestBody);
                    responseData = handleProxyFn(responseBody, mock, {
                        requestQuery: requestQuery,
                        requestBody: requestBody,
                        requestHeaders: requestHeaders
                    }, responseProxyFn);
                    response = mockResponse
                        .status(responseData.responseCode)
                        .body(responseData.responseBody)
                        .headers(responseData.responseHeaders);
                    return [4 /*yield*/, delay(mock.delay)];
                case 1:
                    _a.sent();
                    if (loggingEnabled) {
                        logMock(mock, {
                            requestQuery: requestQuery,
                            requestBody: requestBody,
                            requestHeaders: requestHeaders
                        }, responseData);
                    }
                    return [2 /*return*/, response];
            }
        });
    }); }));
});
