// http-mocks/esm2015/lib/http-mocks.model.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/http-mocks.model.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function MockRequest() { }
if (false) {
    /** @type {?} */
    MockRequest.prototype.requestQuery;
    /** @type {?} */
    MockRequest.prototype.requestBody;
    /** @type {?} */
    MockRequest.prototype.requestHeaders;
}
/**
 * @record
 */
export function MockResponse() { }
if (false) {
    /** @type {?} */
    MockResponse.prototype.responseBody;
    /** @type {?} */
    MockResponse.prototype.responseCode;
    /** @type {?} */
    MockResponse.prototype.responseHeaders;
}
/**
 * @record
 */
export function Mock() { }
if (false) {
    /**
     * A regular expression that should match with the URL of the HTTP request.
     * @type {?}
     */
    Mock.prototype.url;
    /**
     * HTTP method.
     * @type {?}
     */
    Mock.prototype.method;
    /**
     * A function that contains a logic that returns response data accordingly to the request (query params and payload).
     * @type {?}
     */
    Mock.prototype.responseFn;
    /**
     * Status code of the HTTP transaction. (default: `200`)
     * @type {?|undefined}
     */
    Mock.prototype.responseCode;
    /**
     * Headers of the response. (default: `{}`)
     * @type {?|undefined}
     */
    Mock.prototype.responseHeaders;
    /**
     * Time of the response delay (in milliseconds). (default: `0`)
     * @type {?|undefined}
     */
    Mock.prototype.delay;
}
/**
 * @record
 */
export function MockScenarios() { }
if (false) {
    /** @type {?} */
    MockScenarios.prototype.default;
    /* Skipping unhandled member: [mockScenario: string]: Mock[];*/
}
/**
 * @record
 */
export function MockOptions() { }
if (false) {
    /**
     * Whether pass through the request to the network if no mock exists. (default: `true`)
     * @type {?|undefined}
     */
    MockOptions.prototype.fallbackToNetwork;
    /**
     * Whether log information about request/response of the mock to the console. (default: `false`)
     * @type {?|undefined}
     */
    MockOptions.prototype.loggingEnabled;
    /**
     * Whether accept search string placed after the hashtag in URL,
     * e.g. `http://localhost:4200/#/?mockScenario=scenarioKey`. (default: `false`)
     * @type {?|undefined}
     */
    MockOptions.prototype.useLocationHash;
    /**
     * The scenario key to use for mocking. (default: `'default'`)
     * @type {?|undefined}
     */
    MockOptions.prototype.mockScenario;
    /**
     * Common response code for all mocks that can be overridden by the response code of the particular mock. (default: `200`)
     * @type {?|undefined}
     */
    MockOptions.prototype.defaultResponseCode;
    /**
     * Common response headers for all mocks that can be overridden by the response headers of the particular mock. (default: `{}`)
     * @type {?|undefined}
     */
    MockOptions.prototype.defaultResponseHeaders;
    /**
     * Common mock delay for all mocks that can be overridden by the mock delay of the particular mock. (default: `0`)
     * @type {?|undefined}
     */
    MockOptions.prototype.defaultDelay;
    /**
     * Wrapper function that applies to all mock responses.
     * Useful when you change the overall data response structure without an impact on the data.
     * @type {?|undefined}
     */
    MockOptions.prototype.responseProxyFn;
    /**
     * Disables specific mocks.
     * With this option, you can easily disable particular mocks when you're about to connect to the real API.
     * @type {?|undefined}
     */
    MockOptions.prototype.ignoredMocks;
}
